From e34d3383e850d08a0f573b9b50ef13a17c0769ff Mon Sep 17 00:00:00 2001
From: Jiaqing Zhao <jiaqing.zhao@intel.com>
Date: Tue, 31 Oct 2023 06:38:38 +0000
Subject: [PATCH 24/54] Support booting from device not on PCI bridge

Currently kernelflinger assumes the boot device is always behind a
PCI bridge. But the device might be directly on the root complex, like
the passthrough device or emulated device in ACRN, such assumption
causes boot failure.

Fix it by using the BDF directly when boot device is on bus 0.

Tracked-On: OAM-113030
Signed-off-by: Jiaqing Zhao <jiaqing.zhao@intel.com>
---
 libkernelflinger/android.c | 25 +++++++++++++++----------
 1 file changed, 15 insertions(+), 10 deletions(-)

diff --git a/libkernelflinger/android.c b/libkernelflinger/android.c
index 9e4db87..a27e59b 100644
--- a/libkernelflinger/android.c
+++ b/libkernelflinger/android.c
@@ -1375,20 +1375,25 @@ static EFI_STATUS setup_command_line(
 		UINT32 disk_bus_num, storage_bus_num;
 
 		disk_bus_num = get_bootdev_diskbus();
-		debug(L"nvme controller diskbus = %x\n", disk_bus_num);
+		debug(L"diskbus = %x", disk_bus_num);
 
-		storage_bus_num = disk_bus_num>>16;
-
-		bdf = bridge_diskbus(storage_bus_num);
-		if (bdf < 0) {
-			error(L"No pci bridge found, assue bus is 0");
-			bdf = 0;
-		}
+		storage_bus_num = disk_bus_num >> 16;
+                if (storage_bus_num == 0) {
+                        bdf = (disk_bus_num >> 8);
+                } else {
+                        bdf = bridge_diskbus(storage_bus_num);
+                        if (bdf < 0) {
+                                error(L"No pci bridge found");
+                                ret = EFI_INVALID_PARAMETER;
+                                goto out;
+                        }
+                }
 
-                diskbus = PoolPrint(L"%02x.%x", (bdf>>3)&0x1f, bdf&0x7);
+                diskbus = PoolPrint(L"%02x.%x", (bdf >> 3) & 0x1f, bdf & 0x7);
+                debug(L"pci boot device: device = %x func = %x", (bdf >> 3) & 0x1f, bdf & 0x7);
 #else
                 diskbus = PoolPrint(L"%02x.%x", boot_device->Device, boot_device->Function);
-                debug(L"pci bridge: device = %x func = %x \n", boot_device->Device, boot_device->Function);
+                debug(L"pci boot device: device = %x func = %x", boot_device->Device, boot_device->Function);
 #endif
 #else
                 diskbus = PoolPrint(L"%a", (CHAR8 *)PREDEF_DISK_BUS);
-- 
2.25.1

